### DISCUSSION TITLE
Email replyTo

### DISCUSSION THREAD

**Discussion by miztroh** [pocketbase/pocketbase#4595] [Q&A]
Below is the example code for sending emails with go.  Does this mechanism support specifying a replyTo address?  I've not been able to find a way to do this.  This would be highly useful.  Sometimes, we need to send emails via a no-reply mailbox.  If the recipient clicks reply, the follow-up email should go to a different address.  Please advise.  Thank you!

```go
app.OnRecordAfterCreateRequest("users").Add(func(e *core.RecordCreateEvent) error {
        message := &mailer.Message{
            From: mail.Address{
                Address: app.Settings().Meta.SenderAddress,
                Name:    app.Settings().Meta.SenderName,
            },
            To:      []mail.Address{{Address: e.Record.Email()}},
            Subject: "YOUR_SUBJECT...",
            HTML:    "YOUR_HTML_BODY...",
            // bcc, cc, attachments and custom headers are also supported...
        }

        return app.NewMailClient().Send(message)
    })
```

**Comment by ganigeorgiev** [comment]
If changing the From address is not an option, you can try to manually add it to the [mailer.Message](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.4/tools/mailer#Message) Headers field, something like:

```go
message := &mailer.Message{
    Headers: map[string]string{
        "Reply-To": "John Doe <test@example.com>",
    },
    ...
}
```

**Comment by miztroh** [comment]
This does not appear to work.  I receive the email, but the "Reply-To" header doesn't seem to come through and, clicking reply, the address for my reply email is the same as the sending address.

**Reply by ganigeorgiev** [reply]
I've just tested it and the header is added correctly for me:

```
From: Support <support@example.com>
MIME-Version: 1.0
Date: Wed, 20 Mar 2024 08:18:01 +0200
Subject: YOUR_SUBJECT...
To: test@example.com
Reply-To: John Doe <test@example.com>
Message-ID: <fFl7JvICM9X3CT6@example.com>
Content-Type: multipart/mixed;
...
```

Note that it is up to the mail client whether they will acknowledge the header or just ignore it.

**Reply by miztroh** [reply]
OK, that's good to know.  I'll test it with a different email configuration.  Thanks!

**Reply by miztroh** [reply]
@ganigeorgiev I've seen errors like the one below in other go projects.  How can I ignore cert issues with my email server?  I won't have a public cert until deploying for production, so I have to make do with an internal cert for now.

```
Failed to send the test email. Raw error: tls: failed to verify certificate: x509: “redacted.example.com” certificate is not standards compliant.
```

**Reply by ganigeorgiev** [reply]
Is your email server self-hosted/local or some 3rd party service?

If it is the latter and you are in a Docker environment with stripped base image, make sure that you have the root certificates package installed (usually named `ca-certificates`).

If it is a self-hosted mail server with none or custom self-signed certificate, then it is not possible at the moment because we don't expose the [`tls/Config.InsecureSkipVerify`](https://pkg.go.dev/crypto/tls#Config.InsecureSkipVerify) option (_this was requested before in https://github.com/pocketbase/pocketbase/discussions/669 but I'm still not sure that it would be a good idea and I don't want to clutter the interface with every possible knob_).

**Reply by miztroh** [reply]
In my particular case, my employer runs an on-premises SMTP relay backed by Exchange Online.  It is a very common practice and these sorts of SMTP relays almost always use an internal certificate.  For example, organizations run Active Directory Certificate Services in conjunction with Active Directory.  This gives them an internal root / intermediate CA that can be trusted by all computers on the domain.  I've already added the root / intermediate files to my trusted CAs, which works for things like accessing corporate websites.

I think you're going to have to give people the option.  Large mail providers like Exchange Online are looking to [deprecate basic SMTP auth](https://learn.microsoft.com/en-us/exchange/clients-and-mobile-in-exchange-online/deprecation-of-basic-authentication-exchange-online) and so relay servers will become the only option for this sort of custom integration.  I realize it can be a footgun, but people still need the freedom to run this in a way that makes sense for their context.  Is it possible to make the change myself if I'm already extending PocketBase via go?

**Reply by ganigeorgiev** [reply]
It is not possible from the PocketBase config to apply this change.

Based on the [Go source](https://cs.opensource.google/go/go/+/master:src/crypto/x509/root_unix.go;l=16-26) (_at least for unix based platforms_), you can define the env variables `SSL_CERT_FILE`/`SSL_CERT_DIR` pointing to your self-signed cert(s), but I never tried it and I'm not sure if it will work in this case.

**Reply by j0sh** [reply]
@miztroh Wild guess, but do you have the correct internal CA for your SMTP server installed in case it is different from the one used for your websites? Golang should be pulling from the system roots which would include any internal CAs.

**Reply by DasDunkel** [reply]
@ganigeorgiev sorry for jumping in on this, but this option has become **essential** for me to use SMTP at all with pocketbase.

I use proton for my emails, and due to their heavy security I am required to run a relay server which uses a self signed certificate that is only valid for `127.0.0.1`, so it doesn't work on external machines/servers, and since I run pocketbase on an external machine, it doesn't work

Moving services around isn't currently possible, so having the option to disable certificate verification is the only way forward for me, I'd even take an environment variable over a UI toggle if you're worried about a few toggles ruining the UI  (though function over form imho)

**Reply by ganigeorgiev** [reply]
@DasDunkel No sorry, I have no idea what do you mean by _"it doesn't work on external machines/servers"_ but I'm not planning adding the `InsecureSkipVerify` option.

For me the correct approach is to load/install the certificate on the machine where PocketBase is running and not disabling the TLS checks.

**Reply by DasDunkel** [reply]
What I mean is only services running on the same machine as the relay can connect unless verification is disabled

As far as I know installing the certificates in the PocketBase container isn't possible, or at least isn't easy, as Proton Bridge encrypts all of it's data including the certificates (if it doesn't, I certainly can't find them)

**Reply by ganigeorgiev** [reply]
PocketBase (or Go to be more precise) by default uses the system root certificates. Where the CA certificate directory is exactly would depend on the underlying OS but usually it should be enough to copy your `.crt` file into the `/usr/local/share/ca-certificates` directory and rebuild the loaded certificates with `dpkg-reconfigure ca-certificates`/`update-ca-certificates` or similar (_alternatively you can also check the env variables mentioned in https://github.com/pocketbase/pocketbase/discussions/4595#discussioncomment-8855032 to set a different location_). In your case it could be slightly more complicated because it seems that you are also running in a container environment and it would depend on the used image.

> Proton Bridge encrypts all of it's data including the certificates 

Honestly I have no idea what this means but again - for me it is wrong to disable the TLS checks as it makes the self-signed certificate practically pointless.

**Reply by DasDunkel** [reply]
Proton bridge, the relay, encrypts all of it's config/data files to maintain privacy and security, since I can't find a relevant crt file, it's safe to assume it is encrypted too, so it's inaccessible to me

I'd rather not disable certificate authentication, but if I can't get the certificate, and PocketBase can't request and store the certificate itself, it's my best option

**Reply by ganigeorgiev** [reply]
@DasDunkel I'm not familiar with Proton Mail bridge and have no idea how it works but quick google search leads to related SO that seems to contain information about the location of the Proton's certificate https://stackoverflow.com/questions/57746821/getting-a-self-signed-certificate-error-with-protonbridge-and-mbsync#answer-70687229.

In any case, this entire discussion is off-topic to the original question and I'll stop engage further. My recommendation from [my previous reply](https://github.com/pocketbase/pocketbase/discussions/4595#discussioncomment-12841389) remains and I have no plans enabling the `InsecureSkipVerify` option as I don't think it would be the right solution and could introduce unnecessary complexity (_even if we don't add a UI option because once this is added for the email configuration users will start demanding it for the S3 storage, S3 backups, the JSVM `$https.send`, etc. and I don't want to maintain this_).
